### PR TITLE
Slack: Re-use http client and consume response

### PR DESCRIPTION
**Pull request by perttuk**
Re-use http client and consume response.

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackConsumer.java [hunk 1]
```diff
@@ -48,10 +48,12 @@
     private SlackEndpoint slackEndpoint;
     private String timestamp;
     private String channelId;
+    private HttpClient client;
 
     public SlackConsumer(SlackEndpoint endpoint, Processor processor) throws IOException, DeserializationException {
         super(endpoint, processor);
         this.slackEndpoint = endpoint;
+        this.client = HttpClientBuilder.create().useSystemProperties().build();
```
Comment (davsclaus) [inline components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackConsumer.java:56]: Do not create the client in the constructor, but in doStart

File: components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java [hunk 2]
```diff
@@ -32,21 +32,23 @@
...
 public class SlackProducer extends DefaultProducer {
 
-    private SlackEndpoint slackEndpoint;
+    private final SlackEndpoint slackEndpoint;
+    private final HttpClient client;
 
     public SlackProducer(SlackEndpoint endpoint) {
         super(endpoint);
         this.slackEndpoint = endpoint;
+        this.client = HttpClientBuilder.create().useSystemProperties().build();
```
Comment (davsclaus) [inline components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java:45]: Same as consumer

**Review by davsclaus — CHANGES_REQUESTED** [review]
And in doStop then the client should be closed/stopped

**Review by oscerd — CHANGES_REQUESTED** [review]
Is there any particular reason to switch to async? I don't thibk there is a gain in this scenario

**Comment by perttuk** [comment]
> 
> 
> Is there any particular reason to switch to async? I don't thibk there is a gain in this scenario

No, not really. I was thinking it would be minor improvement :-) Should I revert the change?

**Comment by oscerd** [comment]
I was just curious about the idea. It doesn't harm, so no need for reverting

File: components/camel-slack/pom.xml [hunk 3]
```diff
@@ -62,6 +62,10 @@
             <groupId>org.apache.httpcomponents</groupId>
```
Comment (davsclaus) [inline components/camel-slack/pom.xml:62]: Can the old client be removed?

File: components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java [hunk 4]
```diff
@@ -59,7 +61,7 @@ protected void doStop() throws Exception {
     }
```
Comment (davsclaus) [inline components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java:63]: And you have a doStop method where you stop/close the client?

**Review by davsclaus — COMMENTED** [review]
The async is better as its non-blocking

**Review by davsclaus — APPROVED** [review]

File: components/camel-slack/pom.xml (hunk 3 above)
Comment (perttuk) [reply components/camel-slack/pom.xml:62]: It is still used by consumer

File: components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java (hunk 4 above)
Comment (perttuk) [reply components/camel-slack/src/main/java/org/apache/camel/component/slack/SlackProducer.java:63]: I added that in previous commit (https://github.com/apache/camel/pull/4990/commits/424273684ce44708461b0188cc90832517aae26b)

**Pull request merged**